Rate limit thin line

tmessi opened the pull request:
This adds the remaining pieces to complete a thin line for applying rate
limits to the controller api:

1. Adds an http middleware handler function to check and apply the
  limits. This middleware is added to the controller's API handler
  functions to check if a request is allowed by a rate.Limiter prior to
  processing the request.
2. Adds a rate.Limiter to the controller. This is initialized by
   generating a slice of rate.Limits from the ratelimit.Config provided
   on the Controller's config.

I am just opening this as a draft since there are several outstanding
things:

1. I still don't like some of the naming for the access to the action
   registrar. So any suggestions there are welcome.
2. In registering actions for each resource, I discovered a variety of
   ways the actions get defined and used in the app services. See
   b35da8fc2 for more details.  I think this can be cleaned up, but that
   might come later after more of the rate limiting functionality is
   done.
3. Additional tests need to be added in a number of places.

mgaffney reviewed (commented):
Great work!

tmessi commented on internal/ratelimit/config.go:
```diff
@@ -28,3 +41,124 @@ type Config struct {
+	for _, cc := range c {
+		var resourceSet []resource.Type
+		switch {
+		case len(cc.Resources) == 1 && cc.Resources[0] == "*":
```
Yea, and actually we already have some, `resource.All` and `action.All`

tmessi commented on internal/ratelimit/handler.go:
```diff
@@ -0,0 +1,125 @@
+			// The only other error here should be rate.ErrLimitNotFound, which
+			// shouldn't be possible given how we initialize the limiter,
+			// so perhaps a 500 makes sense?
+			rw.WriteHeader(http.StatusInternalServerError)
```
I could see a 503 when the error is `rate.ErrLimiterFull`. I was going to do a 429, but the 503 might be better to distinguish between being rate limited vs the server being overloaded.

However, in the case of `rate.ErrLimitNotFound` I don't think it would be temporary, since it would indicate that the limiter was not properly initialized.

jimlambrt replied:
Hum... thinking a 429 if it's full and a 503 if it's not properly initialized.
Returning a 503 when the service is available but overloaded seems odd to me
since the client should likely behave the same way when experiencing a rate
limit or an overloaded service because it's out of quota.

tmessi replied:
I am not sure I like a `503` for when it is not properly initialized, since it is not something that would be fixed by a configuration change, this would be a bug in the particular release and the only way the service would become available again is with an upgrade/downgrade. This is akin to a panic so I think a `500` would be better.

For when the limiter is full, I do kind of like `503` better. A `429` states:
> The 429 status code indicates that the user has sent too many requests in a given amount of time ("rate limiting").

But when the limiter is full, it is not that the user made too many requests, it is that the service is overwhelmed, which seems more in line with a `503` (my emphasis):
> The 503 (Service Unavailable) status code indicates that the server is currently unable to handle the request **due to a temporary overload** or scheduled maintenance, which will likely be alleviated after some delay.

In addition, with the 503 we can still set a `Retry-After` based on the time until quotas will be deleted, but we would not include the `RateLimit` header, since no quota was created.

As for clients, they could:
1. Minimally check for the `Retry-After` header and retry after that time.
2. Be explicit and check for either a 429 or 503 and the `Retry-After` header.

This would also help operators, since, with our existing monitoring, they could differentiate between requests that consumed too much quota (429), and requests that failed because of the limiter running out of space (503).

tmessi replied:
44d49eb0d switches to a 503 on ErrLimiterFull, on other errors it is still a 500. @jimlambrt let me know if you disagree with my reasoning [above](https://github.com/hashicorp/boundary/pull/3825#discussion_r1376299682), I'm happy to continue discussing it.

jimlambrt replied:
I'm fine with it.  My thinking was: few clients handle 503s as a retry-able error but most will deal with 429s properly.  In other words, a 503 will like result in the client returning an error to the caller and a 429 will likely be retried before returning an error to the caller.

jimlambrt commented on internal/types/action/registrar.go:
```diff
@@ -0,0 +1,70 @@
+	mu sync.RWMutex
+	m  map[resource.Type]*resourceActionSets
+}
+
+func (b *byResource) add(r resource.Type, individual ActionSet, collection ActionSet) error {
```
should this just panic if there's an error since it's only used during
registration which should only happen during init?

tmessi replied:
I updated RegisterResource to panic if called twice with the same resource, see: 21f7ac863. And then I refactored where this is called with the fixup here: 0cb15e6ec.

jimlambrt commented on internal/types/action/registrar.go:
```diff
@@ -0,0 +1,70 @@
+
+// SetForResource returns the valid ActionSet for the given resource.
+// If the provided resource was not registered, and error is returned.
+func SetForResource(r resource.Type) (ActionSet, error) {
```
I like the original `action.SetForResource(...)` over
`action.ResourceActionSet(...)`.  With that said, what about `action.RegisteredActionSet(...)`?

tmessi replied:
To address some TODOs and the comment [here](https://github.com/hashicorp/boundary/pull/3825#discussion_r1373200746), I refactored `action.ActionSet` so it is implemented as a set, see 66e44b75f.

In that commit, I introduce a couple additional package functions that operate on ActionSets. For them I opted to use "ActionSet" in the function name, ie:

- `action.ActionSetUnion`
- `action.ActionSetDifference`

These stutter, which I would generally want to avoid, but it is consistent with `action.ActionSet`. I think I would prefer to either:
1. Be consistent, in which case I think function in question here should be `action.ActionSetForResource`
3. Or fix all of the stutters and refactor `action.ActionSet` to `action.Set`. This would allow this function to be `action.SetForResource` which could still be confusing if you see "Set" and think of a setter. However, since this is a package function rather than a method on a struct, and if there is a type called Set, I don't think it should be that confusing.

jimlambrt commented on internal/ratelimit/handler.go:
```diff
@@ -0,0 +1,125 @@
+	rate "github.com/hashicorp/go-rate"
+)
+
+var pathRegex = regexp.MustCompile(`/v1/(?P<resource>[\w-]+)((/(?P<id>[\w/]+))?(:(?P<action>[\w-:]+)?)?)?`)
```
giddy up!

jimlambrt commented on internal/ratelimit/handler.go:
```diff
@@ -0,0 +1,125 @@
+
+var pathRegex = regexp.MustCompile(`/v1/(?P<resource>[\w-]+)((/(?P<id>[\w/]+))?(:(?P<action>[\w-:]+)?)?)?`)
+
+func extractResourceAction(path, method string) (res, act string, err error) {
```
I sort of like @mgaffney 's suggestion to return defined error codes which can
be used by the caller (the handler) to return the appropriate error/status.

tmessi replied:
See: https://github.com/hashicorp/boundary/pull/3825#discussion_r1376685709

jimlambrt commented on internal/ratelimit/config.go:
```diff
@@ -28,3 +41,124 @@ type Config struct {
+			for _, r := range cc.Resources {
+				rr, ok := resource.Map[r]
+				if !ok {
+					return nil, fmt.Errorf("unknown resource %s", r)
```
should we define a new error code for this?

tmessi replied:
That seems like a good idea. In looking through the existing codes, I am a little surprised that we don't have something for a configuration error, which is what I was thinking this would be. It seems that for most of the config parsing we don't return errors with error codes.

Do you think I should make an error code specific to specifying a resource that is not known? Or should I look to add a more generic "configuration error"?

jimlambrt replied:
I'd start with `configuration error` in this case.

jimlambrt commented on internal/ratelimit/config.go:
```diff
@@ -28,3 +41,124 @@ type Config struct {
+				for _, aStr := range cc.Actions {
+					a, ok := validActionMap[aStr]
+					if !ok {
+						return nil, fmt.Errorf("action %s not valid for resource %s", aStr, res.String())
```
should we define a new error code for this?

tmessi commented on internal/ratelimit/handler.go:
```diff
@@ -0,0 +1,125 @@
+		res, a, err := extractResourceAction(path, method)
+		if err != nil {
+			// TODO: what to do here?
+			// We could:
```
For this and the status codes mentioned in comments above, see 1e1c7f9b6

tmessi commented on internal/daemon/controller/handlers/accounts/account_service.go:
```diff
@@ -108,6 +108,23 @@ func init() {
 		panic(err)
 	}
+
+	// TODO: make action.ActionSet a Set so this isnt necessary?
```
I refactored `action.ActionSet` to be implemented as a set here: 66e44b75f
Then these TODOs are removed here: b5bfde605

jimlambrt commented on internal/daemon/controller/handlers/authmethods/z_init.go:
```diff
@@ -0,0 +1,20 @@
+	"golang.org/x/exp/maps"
+)
+
+// init functions process in alphabetical order of filenames within a package.
```
ty for the comment!

jimlambrt reviewed (commented):
Looks great.  Moving fwd!

tmessi commented:
@mgaffney @jimlambrt I moved this out of draft since it should be ready for a final review. All changes since the last review are done as fixup commits. They are mostly additional tests, some minor fixes due to test failures, and some fixes to handler and api test that has assertions that relied on the order of action sets.

Also, note that `make-gen-detla` is failing, I believe this is unrelated and should be resolved when I rebase the llb on main.

jimlambrt commented on internal/daemon/controller/controller.go:
```diff
@@ -119,6 +120,8 @@ type Controller struct {
 	apiGrpcServerListener grpcServerListener
 	apiGrpcGatewayTicket  string
 
+	ratelimiter *rate.Limiter
```
```suggestion
	rateLimiter *rate.Limiter
```

tmessi replied:
Thanks, fixed in [03ef8dd](https://github.com/hashicorp/boundary/pull/3825/commits/03ef8ddb690b60ba9a01bf9e05021c321e373b53)

jimlambrt commented on internal/types/action/registrar.go:
```diff
@@ -0,0 +1,73 @@
+
+// RegisterResource registers ActionSets for r. If RegisterResource is called
+// twice with the same resource, it panics.
+func RegisterResource(r resource.Type, individual ActionSet, collection ActionSet) {
```
do we want unit tests for these two exported functions.

tmessi replied:
Added a test in [670d49b](https://github.com/hashicorp/boundary/pull/3825/commits/670d49b35466294b14617cd446a6ec05d44ed3e2) and then expanded on the test some in [cda1217](https://github.com/hashicorp/boundary/pull/3825/commits/cda1217130627c6280d87fdd810cf594f7804401). I looked at adding some additional test cases, like attempting to register and invalid `resource.Type`, but it seems we don't have an existing way to checking if a resource.Type is valid currently. I can look into adding some validation logic for it, but I think that should be a separate PR.

tmessi commented on internal/daemon/controller/controller.go:
```diff
@@ -245,6 +248,16 @@ func New(ctx context.Context, conf *Config) (*Controller, error) {
 	}
 	c.clusterListener = clusterListeners[0]
 
+	ratelimits, err := conf.RawConfig.Controller.ApiRateLimits.Limits(c.baseContext)
```
Ty, fixed in [badaf09](https://github.com/hashicorp/boundary/pull/3825/commits/badaf096826502bc289108f8690f6effad122919)

tmessi commented on internal/daemon/controller/handler.go:
```diff
@@ -93,7 +94,8 @@ func (c *Controller) apiHandler(props HandlerProperties) (http.Handler, error) {
 	}
 
-	corsWrappedHandler := wrapHandlerWithCors(mux, props)
+	ratelimitedHandler := ratelimit.Handler(c.ratelimiter, mux)
```
Ty, fixed in [badaf09](https://github.com/hashicorp/boundary/pull/3825/commits/badaf096826502bc289108f8690f6effad122919)

mgaffney commented on internal/ratelimit/config.go:
```diff
@@ -4,15 +4,90 @@
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/types/action"
+	"github.com/hashicorp/boundary/internal/types/resource"
+	rate "github.com/hashicorp/go-rate"
```
```suggestion
	"github.com/hashicorp/go-rate"
```

mgaffney commented on internal/ratelimit/handler.go:
```diff
@@ -0,0 +1,153 @@
+
+	"github.com/hashicorp/boundary/internal/types/action"
+	"github.com/hashicorp/boundary/internal/types/resource"
+	rate "github.com/hashicorp/go-rate"
```
```suggestion
	"github.com/hashicorp/go-rate"
```

mgaffney commented on internal/types/action/action.go:
```diff
@@ -215,29 +215,61 @@ func (act Type) IsActionOrParent(suspect Type) bool {
+}
+
+// ActionSetUnion returns a new ActionSet that is the union sets.
+func ActionSetUnion(sets ...ActionSet) ActionSet {
```
```suggestion
func Union(sets ...ActionSet) ActionSet {
```
Why not just `Union`?

tmessi replied:
For the reasons I outlined [here](https://github.com/hashicorp/boundary/pull/3825#discussion_r1377660801), but happy to reconsider it.

mgaffney commented on internal/types/action/action.go:
```diff
@@ -215,29 +215,61 @@ func (act Type) IsActionOrParent(suspect Type) bool {
+}
+
+// ActionSetDifference returns a new ActionSet that is the difference of a-b.
+func ActionSetDifference(a ActionSet, b ActionSet) ActionSet {
```
```suggestion
func Difference(a ActionSet, b ActionSet) ActionSet {
```
Why not just `Difference`?

mgaffney commented on internal/types/action/action.go:
```diff
@@ -215,29 +215,61 @@ func (act Type) IsActionOrParent(suspect Type) bool {
+	}
+}
+
+// ActionSetUnion returns a new ActionSet that is the union sets.
```
```suggestion
// ActionSetUnion returns a new ActionSet that is the union of the sets.
```

mgaffney commented on internal/daemon/controller/controller.go:
```diff
@@ -50,6 +50,7 @@ import (
 	external_plugins "github.com/hashicorp/boundary/sdk/plugins"
 	"github.com/hashicorp/boundary/version"
 	"github.com/hashicorp/go-hclog"
+	rate "github.com/hashicorp/go-rate"
```
```suggestion
	"github.com/hashicorp/go-rate"
```

jimlambrt reviewed (approved):
ty!
